File: legacy/capacitor-web/src/services/apiClient.js
/**
 * Central backend configuration for the frontend.
 *
 * Use REACT_APP_BACKEND_URL (preferred) to override per environment.
 * Falls back to the deployed backend so builds don't accidentally hit localhost.
 */
export const DEFAULT_BASE_URL = 'https://detea-backend.onrender.com';

export const BASE_URL = (
  process.env.REACT_APP_BACKEND_URL ||
  process.env.REACT_APP_VERTEX_BACKEND_URL ||
  process.env.REACT_APP_VERTEX_GEMINI_URL ||
  DEFAULT_BASE_URL
)
  .trim()
  .replace(/\/$/, '');

/**
 * Fetch JSON from the backend with consistent error handling.
 *
 * - Handles network errors
 * - Handles non-2xx responses (surfacing backend error messages when present)
 * - Handles non-JSON responses
 */
export async function fetchJson(path, options = {}) {
  const p = typeof path === 'string' ? path : '';
  const finalPath = p.startsWith('/') ? p : `/${p}`;
  const url = `${BASE_URL}${finalPath}`;

  const {
    method = 'POST',
    body,
    signal,
    headers = {},
  } = options;

  let res;
  try {
    res = await fetch(url, {
      method,
      headers: { 'Content-Type': 'application/json', ...headers },
      body: body !== undefined ? JSON.stringify(body) : undefined,
      signal,
    });
  } catch (e) {
    const msg = e && e.message ? e.message : String(e);
    throw new Error(`Network error calling backend (${url}): ${msg}`);
  }

  const text = await res.text();
  let data = {};
  try {
    data = text ? JSON.parse(text) : {};
  } catch {
    throw new Error(`Backend returned non-JSON (${res.status}) from ${url}: ${(text || '').slice(0, 240)}`);
  }

  if (!res.ok) {
    const msg = data?.error || data?.message || text || res.statusText || 'Request failed';
    throw new Error(typeof msg === 'string' ? msg : JSON.stringify(msg));
  }

  return data;
}

File: flutter_app/lib/services/api_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../config/env.dart';

/// Mirrors src/services/apiClient.js — always uses [Env.baseUrl] (`BACKEND_URL`).
String get baseUrl => Env.baseUrl;

Future<Map<String, dynamic>> fetchJson(
  String path, {
  String method = 'POST',
  Map<String, dynamic>? body,
  Map<String, String>? headers,
  http.Client? client,
}) async {
  final finalPath = path.startsWith('/') ? path : '/$path';
  final url = Uri.parse('$baseUrl$finalPath');
  final c = client ?? http.Client();

  try {
    late http.Response res;
    final allHeaders = {'Content-Type': 'application/json', ...?headers};

    switch (method.toUpperCase()) {
      case 'GET':
        res = await c.get(url, headers: allHeaders);
        break;
      case 'PUT':
        res = await c.put(url, headers: allHeaders, body: body != null ? jsonEncode(body) : null);
        break;
      case 'DELETE':
        res = await c.delete(url, headers: allHeaders, body: body != null ? jsonEncode(body) : null);
        break;
      default:
        res = await c.post(url, headers: allHeaders, body: body != null ? jsonEncode(body) : null);
    }

    final text = res.body;
    Map<String, dynamic> data = {};
    try {
      data = text.isNotEmpty ? jsonDecode(text) as Map<String, dynamic> : {};
    } catch (_) {
      throw Exception(
        'Backend returned non-JSON (${res.statusCode}) from $url: ${text.length > 240 ? text.substring(0, 240) : text}',
      );
    }

    if (res.statusCode < 200 || res.statusCode >= 300) {
      final msg = data['error'] ?? data['message'] ?? text;
      throw Exception(msg is String ? msg : jsonEncode(msg));
    }

    return data;
  } catch (e) {
    if (e is Exception) rethrow;
    throw Exception('Network error calling backend ($url): $e');
  }
}
